<a href="https://colab.research.google.com/github/argennof/EyCD2022/blob/main/Entregable_2_Nico_parte_1_ejercicio_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import matplotlib.pyplot as plt
import numpy
import pandas as pd

import seaborn
seaborn.set_context('talk')

from sqlalchemy import create_engine, text

Definimos el Dataset "df_melb" con los datos de las viviendas de Melbourne, conjunto de datos de la compentencia Kaggle sobre estimación de precios de ventas de propiedades en Melbourne, Australia.

In [ ]:
url = 'https://cs.famaf.unc.edu.ar/~mteruel/datasets/diplodatos/melb_data.csv'
df_melb = pd.read_csv(url)

Creamos una base de datos en SQLite utilizando la libreria SQLalchemy del dataset df_melb.

In [ ]:
engine = create_engine('sqlite:///melbourne_data.sqlite3', echo=True)

In [ ]:
df_melb.to_sql('melb_data', con=engine, if_exists="replace")

2022-05-26 20:48:25,425 INFO sqlalchemy.engine.Engine PRAGMA main.table_info("melb_data")
2022-05-26 20:48:25,434 INFO sqlalchemy.engine.Engine [raw sql] ()
2022-05-26 20:48:25,437 INFO sqlalchemy.engine.Engine PRAGMA temp.table_info("melb_data")
2022-05-26 20:48:25,439 INFO sqlalchemy.engine.Engine [raw sql] ()
2022-05-26 20:48:25,450 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2022-05-26 20:48:25,454 INFO sqlalchemy.engine.Engine 
CREATE TABLE melb_data (
	"index" BIGINT, 
	"Suburb" TEXT, 
	"Address" TEXT, 
	"Rooms" BIGINT, 
	"Type" TEXT, 
	"Price" FLOAT, 
	"Method" TEXT, 
	"SellerG" TEXT, 
	"Date" TEXT, 
	"Distance" FLOAT, 
	"Postcode" FLOAT, 
	"Bedroom2" FLOAT, 
	"Bathroom" FLOAT, 
	"Car" FLOAT, 
	"Landsize" FLOAT, 
	"BuildingArea" FLOAT, 
	"YearBuilt" FLOAT, 
	"CouncilArea" TEXT, 
	"Lattitude" FLOAT, 
	"Longtitude" FLOAT, 
	"Regionname" TEXT, 
	"Propertycount" FLOAT
)


2022-05-26 20:48:25,457 INFO sqlalchemy.engine.Engine [no key 0.00315s] ()
2022-05-26 20:48:25,495 INFO sqlal

Gerenamos un nuevo Dataset "df_airbnb" con las columnas de interes vistas en clases.

In [ ]:
interesting_cols = [
  'description', 'neighborhood_overview',
  'street', 'neighborhood', 'city', 'suburb', 'state', 'zipcode',
  'price', 'weekly_price', 'monthly_price',
  'latitude', 'longitude',
]
df_airbnb = pd.read_csv(
    'https://cs.famaf.unc.edu.ar/~mteruel/datasets/diplodatos/cleansed_listings_dec18.csv',
    usecols=interesting_cols)

/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:2882: DtypeWarning: Columns (35) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


Creamos una base de datos en SQLite utilizando la libreria SQLalchemy del dataset df_airbnb.

In [ ]:
df_airbnb.to_sql('melb_data_airbnb', con=engine, if_exists="replace")

2022-05-26 20:50:53,472 INFO sqlalchemy.engine.Engine PRAGMA main.table_info("melb_data_airbnb")
2022-05-26 20:50:53,474 INFO sqlalchemy.engine.Engine [raw sql] ()
2022-05-26 20:50:53,477 INFO sqlalchemy.engine.Engine PRAGMA temp.table_info("melb_data_airbnb")
2022-05-26 20:50:53,479 INFO sqlalchemy.engine.Engine [raw sql] ()
2022-05-26 20:50:53,489 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2022-05-26 20:50:53,496 INFO sqlalchemy.engine.Engine 
CREATE TABLE melb_data_airbnb (
	"index" BIGINT, 
	description TEXT, 
	neighborhood_overview TEXT, 
	street TEXT, 
	neighborhood TEXT, 
	city TEXT, 
	suburb TEXT, 
	state TEXT, 
	zipcode TEXT, 
	latitude FLOAT, 
	longitude FLOAT, 
	price BIGINT, 
	weekly_price FLOAT, 
	monthly_price FLOAT
)


2022-05-26 20:50:53,499 INFO sqlalchemy.engine.Engine [no key 0.00369s] ()
2022-05-26 20:50:53,517 INFO sqlalchemy.engine.Engine CREATE INDEX ix_melb_data_airbnb_index ON melb_data_airbnb ("index")
2022-05-26 20:50:53,519 INFO sqlalchemy.engine.Engine 

Revisamos las columnas de cada df para poder elegir las variables con las que trabajaremos

In [ ]:
print(df_melb.columns)
print(df_airbnb.columns)

Index(['Suburb', 'Address', 'Rooms', 'Type', 'Price', 'Method', 'SellerG',
       'Date', 'Distance', 'Postcode', 'Bedroom2', 'Bathroom', 'Car',
       'Landsize', 'BuildingArea', 'YearBuilt', 'CouncilArea', 'Lattitude',
       'Longtitude', 'Regionname', 'Propertycount'],
      dtype='object')
Index(['description', 'neighborhood_overview', 'street', 'neighborhood',
       'city', 'suburb', 'state', 'zipcode', 'latitude', 'longitude', 'price',
       'weekly_price', 'monthly_price'],
      dtype='object')


*Generamos* 3 consultas:
> 1- Cantidad de registros totales por barrio de la tabla melb_data<br>
> 2- Cantidad de registros totales por ciudad de la tabla melb_data_airbnb<br>
> 3- Cantidad de registros totales por barrio de la tabla melb_data_airbnb

In [ ]:
query1 = "SELECT Suburb, COUNT(1) as Total_records_Suburb FROM melb_data GROUP BY Suburb ORDER BY Total_records_Suburb DESC"
query2 = "SELECT city as City, COUNT(1) as Total_records FROM melb_data_airbnb GROUP BY city ORDER BY Total_records DESC"
query3 = "SELECT suburb as Suburb, COUNT(1) as Total_records_Suburb_airbnb FROM melb_data_airbnb GROUP BY Suburb ORDER BY Total_records_Suburb_airbnb DESC"



Probamos las consultas<br>


In [ ]:
with engine.connect() as con:
      rs = con.execute(query2)
      print(query2)
      for row in rs:
          print(row)

      print('\n\n')

2022-05-26 20:54:24,384 INFO sqlalchemy.engine.Engine SELECT city as City, COUNT(1) as Total_records FROM melb_data_airbnb GROUP BY city ORDER BY Total_records DESC
2022-05-26 20:54:24,388 INFO sqlalchemy.engine.Engine [raw sql] ()
SELECT city as City, COUNT(1) as Total_records FROM melb_data_airbnb GROUP BY city ORDER BY Total_records DESC
('Melbourne', 7368)
('Port Phillip', 2808)
('Yarra', 2049)
('Stonnington', 1621)
('Moreland', 967)
('Yarra Ranges', 771)
('Darebin', 698)
('Boroondara', 664)
('Glen Eira', 631)
('Whitehorse', 614)
('Monash', 571)
('Maribyrnong', 436)
('Wyndham', 426)
('Bayside', 375)
('Moonee Valley', 344)
('Manningham', 313)
('Kingston', 309)
('Hobsons Bay', 239)
('Banyule', 203)
('Frankston', 177)
('Knox', 175)
('Hume', 170)
('Casey', 153)
('Greater Dandenong', 147)
('Whittlesea', 137)
('Cardinia', 123)
('Maroondah', 115)
('Brimbank', 108)
('Melton', 95)
('Nillumbik', 88)





Controlamos que funcione el código comparando la misma búsqueda con el df "df_airbnb"

In [ ]:
df_airbnb.city.value_counts()

Melbourne            7368
Port Phillip         2808
Yarra                2049
Stonnington          1621
Moreland              967
Yarra Ranges          771
Darebin               698
Boroondara            664
Glen Eira             631
Whitehorse            614
Monash                571
Maribyrnong           436
Wyndham               426
Bayside               375
Moonee Valley         344
Manningham            313
Kingston              309
Hobsons Bay           239
Banyule               203
Frankston             177
Knox                  175
Hume                  170
Casey                 153
Greater Dandenong     147
Whittlesea            137
Cardinia              123
Maroondah             115
Brimbank              108
Melton                 95
Nillumbik              88
Name: city, dtype: int64

Como funciona, ejecutamos las dos queries que faltan.

In [ ]:
with engine.connect() as con:
      rs = con.execute(query3)
      print(query3)
      for row in rs:
          print(row)

      print('\n\n')

2022-05-26 20:55:01,698 INFO sqlalchemy.engine.Engine SELECT suburb as Suburb, COUNT(1) as Total_records_Suburb_airbnb FROM melb_data_airbnb GROUP BY Suburb ORDER BY Total_records_Suburb_airbnb DESC
2022-05-26 20:55:01,701 INFO sqlalchemy.engine.Engine [raw sql] ()
SELECT suburb as Suburb, COUNT(1) as Total_records_Suburb_airbnb FROM melb_data_airbnb GROUP BY Suburb ORDER BY Total_records_Suburb_airbnb DESC
('Melbourne', 4274)
('Southbank', 1236)
('South Yarra', 883)
('Saint Kilda', 805)
('Docklands', 612)
('Richmond', 570)
('Carlton', 537)
('Brunswick', 439)
('Fitzroy', 378)
('North Melbourne', 340)
('Elwood', 328)
('Collingwood', 295)
('South Melbourne', 288)
('Northcote', 283)
('West Melbourne', 263)
('Prahran', 262)
('Abbotsford', 258)
('Port Melbourne', 258)
('Fitzroy North', 238)
('St Kilda', 197)
('Windsor', 195)
('East Melbourne', 193)
('Point Cook', 191)
('Box Hill', 182)
('Hawthorn', 182)
('Saint Kilda East', 182)
('Brunswick East', 179)
('Glen Waverley', 174)
('Footscray', 1

In [ ]:
with engine.connect() as con:
      rs = con.execute(query1)
      print(query1)
      for row in rs:
          print(row)

      print('\n\n')

2022-05-26 20:55:19,512 INFO sqlalchemy.engine.Engine SELECT Suburb, COUNT(1) as Total_records_Suburb FROM melb_data GROUP BY Suburb ORDER BY Total_records_Suburb DESC
2022-05-26 20:55:19,516 INFO sqlalchemy.engine.Engine [raw sql] ()
SELECT Suburb, COUNT(1) as Total_records_Suburb FROM melb_data GROUP BY Suburb ORDER BY Total_records_Suburb DESC
('Reservoir', 359)
('Richmond', 260)
('Bentleigh East', 249)
('Preston', 239)
('Brunswick', 222)
('Essendon', 220)
('South Yarra', 202)
('Glen Iris', 195)
('Hawthorn', 191)
('Coburg', 190)
('Northcote', 188)
('Brighton', 186)
('Kew', 177)
('Balwyn North', 171)
('Pascoe Vale', 171)
('Yarraville', 164)
('St Kilda', 162)
('Glenroy', 159)
('Port Melbourne', 153)
('Moonee Ponds', 149)
('Carnegie', 146)
('Bentleigh', 139)
('Thornbury', 139)
('Brighton East', 138)
('Newport', 137)
('Camberwell', 132)
('Footscray', 132)
('Elwood', 131)
('Ascot Vale', 130)
('Hampton', 130)
('Surrey Hills', 125)
('Prahran', 123)
('Keilor East', 121)
('Hawthorn East', 11

Vamos a ver el contenido de "zipcode" de la tabla melb_data_airbnb para corroborar que coincidan con los valores de "Postcode" de melb_data con la que haremos el JOIN

In [ ]:
query4 = "SELECT zipcode, COUNT(1) as Records_of_zipcodes FROM melb_data_airbnb GROUP BY zipcode ORDER BY Records_of_zipcodes DESC "
with engine.connect() as con:
      rs = con.execute(query4)
      print(query4)
      for row in rs:
          print(row)

      print('\n\n')

2022-05-24 04:13:26,831 INFO sqlalchemy.engine.Engine SELECT zipcode, COUNT(1) as Records_of_zipcodes FROM melb_data_airbnb GROUP BY zipcode ORDER BY Records_of_zipcodes DESC 
2022-05-24 04:13:26,834 INFO sqlalchemy.engine.Engine [raw sql] ()
SELECT zipcode, COUNT(1) as Records_of_zipcodes FROM melb_data_airbnb GROUP BY zipcode ORDER BY Records_of_zipcodes DESC 
('3000.0', 2491)
('3000', 876)
('3006.0', 860)
('3182.0', 579)
('3182', 556)
('3004.0', 555)
('3141.0', 526)
('3008.0', 506)
('3006', 408)
('3121.0', 371)
('3141', 355)
('3053.0', 349)
('3181.0', 275)
('3056.0', 257)
('3121', 257)
('3065', 220)
('3003.0', 216)
('3051.0', 198)
('3030.0', 190)
('3053', 189)
('3056', 189)
('3205.0', 186)
('3181', 185)
('3183.0', 181)
('3184', 181)
('3004', 173)
('3128.0', 169)
('3065.0', 167)
('3066.0', 166)
('3070.0', 164)
('3068', 160)
('3067.0', 157)
('3184.0', 154)
('3068.0', 148)
('3207.0', 146)
('3183', 145)
(None, 142)
('3150.0', 141)
('3051', 140)
('3066', 131)
('3011.0', 128)
('3108.0', 1

Se observan que los dos resultados con mayor frecuencia son 3000.0 con 2491 apariciones y 3000 con 876, siendo en realidad el mismo código postal.<br>
Además, filtramos por zipcode y calculamos el promedio de los precios diarios, semanales y mensuales.

In [ ]:
query5 = """
SELECT zipcode,  
    AVG(price) as price_mean, 
    AVG(weekly_price) as weekly_price_mean, 
    AVG(monthly_price) as monthly_price_mean
FROM (
    SELECT 
        CAST(zipcode AS float) as zipcode,
        price,
        weekly_price,
        monthly_price
    FROM melb_data_airbnb
      )
WHERE zipcode IS NOT NULL
GROUP BY zipcode
ORDER BY zipcode DESC
"""
with engine.connect() as con:
      rs = con.execute(query5)
      print(query5)
      for row in rs:
          print(row)

2022-05-24 04:13:38,039 INFO sqlalchemy.engine.Engine 
SELECT zipcode,  
    AVG(price) as price_mean, 
    AVG(weekly_price) as weekly_price_mean, 
    AVG(monthly_price) as monthly_price_mean
FROM (
    SELECT 
        CAST(zipcode AS float) as zipcode,
        price,
        weekly_price,
        monthly_price
    FROM melb_data_airbnb
      )
WHERE zipcode IS NOT NULL
GROUP BY zipcode
ORDER BY zipcode DESC

2022-05-24 04:13:38,042 INFO sqlalchemy.engine.Engine [raw sql] ()

SELECT zipcode,  
    AVG(price) as price_mean, 
    AVG(weekly_price) as weekly_price_mean, 
    AVG(monthly_price) as monthly_price_mean
FROM (
    SELECT 
        CAST(zipcode AS float) as zipcode,
        price,
        weekly_price,
        monthly_price
    FROM melb_data_airbnb
      )
WHERE zipcode IS NOT NULL
GROUP BY zipcode
ORDER BY zipcode DESC

(30122.0, 75.0, 525.0, 2281.0)
(9540.0, 245.0, None, None)
(9004.0, 226.0, None, None)
(8938.0, 296.0, None, None)
(7367.0, 230.0, None, None)
(7266.0, 220.0

Ahora aplicamos esta consulta pero a la unión de las dos tablas

In [ ]:
query6 = """
SELECT * FROM melb_data
LEFT JOIN (
  SELECT zipcode ,  
    AVG(price) as price_mean, 
    AVG(weekly_price) as weekly_price_mean, 
    AVG(monthly_price) as monthly_price_mean
  FROM (
    SELECT 
        CAST(zipcode AS float) as zipcode,
        price,
        weekly_price,
        monthly_price
    FROM melb_data_airbnb
        )
  WHERE zipcode IS NOT NULL
  GROUP BY zipcode
  ORDER BY zipcode DESC
          )
      ON melb_data_airbnb.zipcode = melb_data.Postcode
"""
with engine.connect() as con:
      rs = con.execute(query6)
      print(query6)
      for row in rs:
          print(row)

2022-05-26 21:16:24,786 INFO sqlalchemy.engine.Engine 
SELECT * FROM melb_data
LEFT JOIN (
  SELECT zipcode ,  
    AVG(price) as price_mean, 
    AVG(weekly_price) as weekly_price_mean, 
    AVG(monthly_price) as monthly_price_mean
  FROM (
    SELECT 
        CAST(zipcode AS float) as zipcode,
        price,
        weekly_price,
        monthly_price
    FROM melb_data_airbnb
        )
  WHERE zipcode IS NOT NULL
  GROUP BY zipcode
  ORDER BY zipcode DESC
          )
      ON melb_data_airbnb.zipcode = melb_data.Postcode

2022-05-26 21:16:24,790 INFO sqlalchemy.engine.Engine [raw sql] ()
2022-05-26 21:16:24,796 INFO sqlalchemy.engine.Engine ROLLBACK


OperationalError: ignored

In [ ]:
df_airbnb.columns.to_list()

['description',
 'neighborhood_overview',
 'street',
 'neighborhood',
 'city',
 'suburb',
 'state',
 'zipcode',
 'latitude',
 'longitude',
 'price',
 'weekly_price',
 'monthly_price']

In [ ]:
query= "SELECT zipcode FROM melb_data_airbnb"
with engine.connect() as con:
      rs = con.execute(query)
      print(query)
      for row in rs:
          print(row)
          break

2022-05-26 21:21:34,286 INFO sqlalchemy.engine.Engine SELECT zipcode FROM melb_data_airbnb
2022-05-26 21:21:34,288 INFO sqlalchemy.engine.Engine [raw sql] ()
SELECT zipcode FROM melb_data_airbnb
('3105',)


Este último punto no se puede visualizar por el OperationalError: RIGHT and FULL OUTER JOINs are not currently supported.